alacritty/src/renderer/rect.rs

```diff
@@ -0,0 +1,259 @@
+static RECT_SHADER_V_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.v.glsl");
+static RECT_SHADER_F: &str = include_str!("../../res/rect.f.glsl");
+static RECT_SHADER_V: &str = include_str!("../../res/rect.v.glsl");
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Rgba {
+    r: u8,
+    g: u8,
+    b: u8,
+    a: u8,
+}
+
+/// Vertex to store solid-color rectangle data.
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Vertex {
+    // Normalized screen coordinates.
+    // TODO these can certainly be i16.
+    x: f32,
+    y: f32,
+
+    // Color.
+    color: Rgba,
+}
+
+/// Structure to store and group together rect-related vertices data and GL objects.
+#[derive(Debug)]
+pub struct RectRenderer {
+    // GL buffer objects. VAO stores attribute and ebo bindings.
+    vao: GLuint,
+    vbo: GLuint,
+    ebo: GLuint,
+
+    program: RectShaderProgram,
+
+    // Accumulated vertices data. Allocated space reused between frames.
+    vertices: Vec<Vertex>,
+}
+
+impl RectRenderer {
+    /// Update program when doing live-shader-reload.
+    pub fn set_program(&mut self, program: RectShaderProgram) {
+        self.program = program;
+    }
+
+    pub fn new() -> Result<Self, Error> {
+        let mut vao: GLuint = 0;
+        let mut vbo: GLuint = 0;
+        let mut ebo: GLuint = 0;
+        let program = RectShaderProgram::new()?;
+
+        // Pre-generate the entire index buffer for max possible rects.
+        let mut indices = Vec::<u16>::new();
+        let max_rects = MAX_VERTICES / 4;
+        indices.reserve(max_rects * 6);
+        for index in 0..max_rects {
+            let index = index as u16 * 4;
+            indices.push(index);
+            indices.push(index + 1);
+            indices.push(index + 2);
+            indices.push(index + 2);
+            indices.push(index + 3);
+            indices.push(index + 1);
+        }
+
+        unsafe {
+            // Allocate buffers.
+            gl::GenVertexArrays(1, &mut vao);
+            gl::GenBuffers(1, &mut vbo);
+            gl::GenBuffers(1, &mut ebo);
+
+            // Bind EBO to VAO.
+            gl::BindVertexArray(vao);
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, ebo);
+
+            // Pre-upload index buffer for max possible rects.
+            gl::BufferData(
+                gl::ELEMENT_ARRAY_BUFFER,
+                (indices.len() * std::mem::size_of::<u16>()) as isize,
+                indices.as_ptr() as *const _,
+                gl::STATIC_DRAW,
+            );
+
+            // VBO binding is not part ot VAO, but attributes are.
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+
+            let mut index = 0;
+            let mut size = 0;
+
+            macro_rules! add_attr {
+                ($count:expr, $gl_type:expr, $type:ty) => {
+                    gl::VertexAttribPointer(
+                        index,
+                        $count,
+                        $gl_type,
+                        gl::FALSE,
```

Comment:
You actually need to pass gl::TRUE for normalized in `color` case, since otherwise it's always white or similar.

Reply:
KEKW true